### PR TITLE
[CI] [GHA] Introduce conformance tests

### PR DESCRIPTION
**Pull request by akashchi**
### Details:
This PR adds the conformance tests into the Linux GHA workflows as a separate job.

### Tickets:
 - *120617*

### CODE REVIEW

File: .github/workflows/linux_conformance.yml [hunk 1]
```diff
@@ -0,0 +1,942 @@
+name: Conformance Tests on Linux (Ubuntu 20.04, Python 3.11)
```
Comment (ilya-lavrenov) [inline .github/workflows/linux_conformance.yml:1]: I hope we will run conformance as a part of main Linux pipeline, right?

Comment (akashchi) [reply .github/workflows/linux_conformance.yml:1]: Yes. I've excluded it into a separate workflow for testing purposes.

File: .github/workflows/linux_conformance.yml [hunk 2]
```diff
@@ -0,0 +1,952 @@
...
+        with:
+          repository: 'openvinotoolkit/openvino_contrib'
+          path: 'openvino_contrib'
+          submodules: 'recursive'
+
+      - name: Clone testdata for C API tests
+        uses: actions/checkout@v4
+        with:
+          repository: 'openvinotoolkit/testdata'
+          path: 'testdata'
```
Comment (iefode) [inline .github/workflows/linux_conformance.yml:64]: we can do not download `testdata` for conformance tests

File: .github/workflows/linux_conformance.yml [hunk 3]
```diff
@@ -0,0 +1,952 @@
...
+          python-version: '3.11'
+
+      - name: Install python dependencies
+        run: |
+          # For Python API
+          python3 -m pip install --upgrade pip
+          python3 -m pip install Scons
+          python3 -m pip install -r ${{ env.OPENVINO_REPO }}/src/bindings/python/wheel/requirements-dev.txt
+          python3 -m pip install -r ${{ env.OPENVINO_REPO }}/src/bindings/python/requirements.txt
+          python3 -m pip install -r ${{ env.OPENVINO_REPO }}/src/bindings/python/requirements_test.txt
```
Comment (iefode) [inline .github/workflows/linux_conformance.yml:93]: `python3 -m pip install -r ${{ env.OPENVINO_REPO }}/src/tests/test_utils/functional_test_utils/layer_test_summary/requirements.txt`

File: .github/workflows/linux_conformance.yml [hunk 4]
```diff
@@ -0,0 +1,952 @@
...
+            -S ${{ env.OPENVINO_REPO }} \
+            -B ${{ env.BUILD_DIR }}
+
+      - name: Clean ccache stats
+        run: ccache --zero-stats --show-config
+
+      - name: Build Conformance Targets
+        working-directory: ${{ env.BUILD_DIR }}
+        run:
+          make --jobs=$(nproc --all) subgraphsDumper
```
Comment (iefode) [inline .github/workflows/linux_conformance.yml]: I believe,`subgraphsDumper` is not required for pre-commit because we will use ready `Conformance IR cache`

File: .github/workflows/linux_conformance.yml [hunk 5]
```diff
@@ -0,0 +1,952 @@
...
+          popd
+
+      - name: Check Dir 1
+        run: ls -laR ${{ env.BUILD_DIR }}
+
+      - name: Check Dir 2
+        run: ls -laR ${{ env.INSTALL_DIR }}
+
+      - name: Check Dir 3
+        run: ls -laR
```
Comment (iefode) [inline .github/workflows/linux_conformance.yml:231]: I do not see any runs.. Am I correct?

Comment (akashchi) [reply .github/workflows/linux_conformance.yml:231]: It is still in the very early stages, I will re-request review once it is in a more ready state.

Comment (iefode) [reply .github/workflows/linux_conformance.yml]: Please, contact me to understand details

File: src/tests/test_utils/functional_test_utils/layer_tests_summary/run_conformance.py [hunk 6]
```diff
@@ -195,14 +208,20 @@ def __check_expected_failures(self):
         diff = this_run_failures.difference(self._expected_failures)
         if len(diff) > 0:
             logger.error(f"Unexpected failures: {diff}")
-            exit(-1)
-
-        intersection = self._expected_failures.intersection(this_run_failures)
-        if this_run_failures != self._expected_failures and self._expected_failures_update:
-            logger.info(f"Expected failures file {self._expected_failures} will be updated!!!")
+            self._unexpected_failures = diff
+            self.is_successful_run = False
```
Comment (iefode) [inline src/tests/test_utils/functional_test_utils/layer_tests_summary/run_conformance.py:212]: return? exit(-1)? Not to do the next step?

Comment (iefode) [inline src/tests/test_utils/functional_test_utils/layer_tests_summary/run_conformance.py:212]: Print `self._unexpected_failures` diff in all cases?

File: src/tests/test_utils/functional_test_utils/layer_tests_summary/run_parallel.py [hunk 7]
```diff
@@ -450,7 +450,7 @@ def __get_filters(self):
             self._is_save_cache = False
             cached_test_list = self.__prepare_smart_filters(cached_test_dict)
         runtime_test_list = list()
-        if len(runtime_test_dist) > 0:
+        if len(runtime_test_dist) > 0 and self._split_unit == constants.SUITE_UNIT_NAME:
```
Comment (iefode) [inline src/tests/test_utils/functional_test_utils/layer_tests_summary/run_parallel.py:453]: Cache is also using for functional tests where suite is base split unit. Should we remove `and self._split_unit == constants.SUITE_UNIT_NAME`? In other case please explain the details

**Review by iefode — APPROVED** [review]

File: .github/workflows/linux.yml [hunk 8]
```diff
@@ -384,6 +388,101 @@ jobs:
...
+          pushd ${INSTALL_DIR}
+            tar -xzf openvino_package.tar.gz -C ${INSTALL_DIR}
+          popd
+          pushd ${INSTALL_TEST_DIR}
+            tar -xzf openvino_tests.tar.gz -C ${INSTALL_DIR}
+          popd
+
+      - uses: actions/setup-python@v4
+        with:
+          python-version: '3.11'
```
Comment (ilya-lavrenov) [inline .github/workflows/linux.yml:446]:
```suggestion
          python-version: ${{ env.PYTHON_VERSION }}
```

File: .github/workflows/linux.yml [hunk 9]
```diff
@@ -384,6 +388,101 @@ jobs:
...
+
+      - name: Install Dependencies
+        run: |
+          sudo -E ${INSTALL_DIR}/install_dependencies/install_openvino_dependencies.sh -c=core -c=dev -y
```
Comment (ilya-lavrenov) [inline .github/workflows/linux.yml:450]:
```suggestion
          ${INSTALL_DIR}/install_dependencies/install_openvino_dependencies.sh -c=core -y
```

**Review by ilya-lavrenov — APPROVED** [review]

File: .github/workflows/linux.yml [hunk 10]
```diff
@@ -384,6 +388,101 @@ jobs:
           path: ${{ env.INSTALL_TEST_DIR }}/TEST*.xml
           if-no-files-found: 'error'
 
+  Conformance_Tests:
```
Comment (ilya-lavrenov) [inline .github/workflows/linux.yml:391]:
```suggestion
  Conformance:
```

File: .github/workflows/linux.yml [hunk 11]
```diff
@@ -384,6 +388,101 @@ jobs:
...
+      run:
+        shell: bash
+    runs-on: ubuntu-20.04-8-cores
+    strategy:
+      max-parallel: 2
+      fail-fast: false
+      matrix:
+        include:
+          # 'OP' for Opset, 'API' for API
+          - TEST_TYPE: 'OP'
```
Comment (ilya-lavrenov) [inline .github/workflows/linux.yml:403]:
```suggestion
          - TEST_TYPE: 'Operations'
```

File: .github/workflows/linux.yml (hunk 9 above)
Comment (akashchi) [reply .github/workflows/linux.yml]: `sudo` is needed as it is run in the GHA runner. Removed `-c=dev`.

File: .github/workflows/linux.yml (hunk 11 above)
Comment (akashchi) [reply .github/workflows/linux.yml:395]: I suggest keeping it as it is because the `run_conformance.py` script accepts `OP` and I do not want to make it uglier with something like `${{ matrix.TEST_TYPE == 'Operations' && 'OP' }}`.

File: src/tests/test_utils/functional_test_utils/layer_tests_summary/run_parallel.py [hunk 12]
```diff
@@ -330,7 +330,7 @@ def __get_test_list_by_runtime(self, test_unit = constants.TEST_UNIT_NAME):
                     if constants.DISABLED_PREFIX in real_test_name:
                         self._disabled_tests.append(real_test_name)
                     elif test_unit == constants.TEST_UNIT_NAME:
-                        tests_dict[real_test_name] = 1
+                        tests_dict[real_test_name] = -1
```
Comment (mryzhov) [inline src/tests/test_utils/functional_test_utils/layer_tests_summary/run_parallel.py:333]: what is the purpose of this change? Those change needed to speedup cache creation.

**Review by ilya-lavrenov — APPROVED** [review]

File: .github/workflows/linux.yml [hunk 13]
```diff
@@ -384,6 +380,101 @@ jobs:
...
+      INSTALL_TEST_DIR: ${{ github.workspace }}/install/tests
+      CONFORMANCE_TOOLS_DIR: ${{ github.workspace }}/install/tests/functional_test_utils/layer_tests_summary
+      CONFORMANCE_ARTIFACTS_DIR: ${{ github.workspace }}/install/conformance_artifacts
+      TEST_DEVICE: 'CPU'
+
+    steps:
+
+      - name: Create Directories
+        run: |
+          mkdir -p ${INSTALL_DIR} ${INSTALL_TEST_DIR} ${CONFORMANCE_ARTIFACTS_DIR}
```
Comment (mryzhov) [inline .github/workflows/linux.yml:409]: INSTALL_DIR and INSTALL_TEST_DIR will be created by Download action (no need to create them separately)

**Review by mryzhov — CHANGES_REQUESTED** [review]

File: src/tests/test_utils/functional_test_utils/layer_tests_summary/run_parallel.py (hunk 12 above)
Comment (akashchi) [reply src/tests/test_utils/functional_test_utils/layer_tests_summary/run_parallel.py:333]: @iefode, could you please elaborate on that change? I'm afraid that I don't remember the exact reasoning.

Comment (iefode) [reply src/tests/test_utils/functional_test_utils/layer_tests_summary/run_parallel.py:333]: @mryzhov You have set the possible execution time. It means that tests is started to be united to processes (but it is mistake). It is not expected by default case. The default approach for `Test` split unit (without cache) is one test per one process. -1 provides the correct distribution.

Comment (mryzhov) [reply src/tests/test_utils/functional_test_utils/layer_tests_summary/run_parallel.py:333]: Why it is mistake? If the test have never been executed (it is not in the cache) why we should run in in separate process?

Comment (iefode) [reply src/tests/test_utils/functional_test_utils/layer_tests_summary/run_parallel.py:333]: @mryzhov I have shared the details how script should work offline and the above comment. I don't understand your point.

Comment (mryzhov) [reply src/tests/test_utils/functional_test_utils/layer_tests_summary/run_parallel.py:333]: never mind if it is already merged
